plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergMetadata.java

```diff
@@ -969,6 +993,83 @@ private void expireSnapshots(Table table, long expireTimestamp, ConnectorSession
                 .collect(toImmutableSet());
     }
 
+    public void executeDeleteOrphanFiles(ConnectorSession session, IcebergTableExecuteHandle executeHandle)
+    {
+        IcebergDeleteOrphanFilesHandle deleteOrphanFilesHandle = (IcebergDeleteOrphanFilesHandle) executeHandle.getProcedureHandle();
+
+        Table table = catalog.loadTable(session, executeHandle.getSchemaTableName());
+        Duration retention = requireNonNull(deleteOrphanFilesHandle.getRetentionThreshold(), "retention is null");
+        validateTableExecuteParameters(
+                table,
+                executeHandle.getSchemaTableName(),
+                "DELETE_ORPHAN_FILES",
+                retention,
+                getDeleteOrphanFilesMinRetention(session),
+                IcebergConfig.DELETE_ORPHAN_FILES_MIN_RETENTION,
+                IcebergSessionProperties.DELETE_ORPHAN_FILES_MIN_RETENTION);
+
+        long expireTimestampMillis = session.getStart().toEpochMilli() - retention.toMillis();
+        deleteOrphanFiles(table, session, executeHandle.getSchemaTableName(), expireTimestampMillis);
+        deleteOrphanMetadataFiles(table, session, executeHandle.getSchemaTableName(), expireTimestampMillis);
+    }
+
+    private void deleteOrphanFiles(Table table, ConnectorSession session, SchemaTableName schemaTableName, long expireTimestamp)
+    {
+        Set<String> validDataFilePaths = stream(table.snapshots())
+                .map(Snapshot::snapshotId)
+                .flatMap(snapshotId -> stream(table.newScan().useSnapshot(snapshotId).planFiles()))
+                .map(fileScanTask -> URI.create(fileScanTask.file().path().toString()).getPath())
```

Comment:
URI.getPath strips information (See earlier comment)

Reply:
Add a comment why we decided to use `URI.getPath` only here. Perhaps a reference to iceberg issue.